v1.7 backport 2020 04 24

jrfastab opened the pull request:
While doing backports into v1.7 we hit a handful of verification errors that related to master branch running clang-11 where v1.7 branch is running clang-7. These were non-trivial to fix without moving code around and creating volatile types, marking a few things READ_ONCE/WRITE_ONCE, etc.

At this point it seems easier to update clang used in v1.7 to clang-10 (a stable clang version) then to create subtle divergent branches. We borrow Dockerfile.runtime from master with some modifications to use clang-10 instead of specific commit. Then bpf code drops stdio.h, stddef,h and stdbool.h usage and instead uses our bpf stddef.h which only creates defines needed by BPF code.

As an example of a verifier error generated on clang-7,

1530: (07) r1 += 8"
1531: (47) r2 |= 3840"
1532: (63) *(u32 *)(r1 +0) = r2"

resulting in a verifier error "dereference of modified ctx ptr R1 off=8 disallowed" clang-10 however generates the code,

r3 |= r2
*(u32 *)(r1 + 8) = r3

where the ctx pointer r1 is not modified by the addition and instead the +8 is included in the load.

Hopefully after this going forward we will get code generation closer between the branches.

pchaigno committed:
bpf: Fix pointer-to-int-cast warning in newer Clang

[ upstream commit e43c2fff3749614a0d5908df6f15db6ea96aed94 ]

Latest Clang (9658d895 "[Sema] Adds the pointer-to-int-cast diagnostic")
emits the following warning for our __fetch code:

    In file included from bpf_lxc.c:18:
    In file included from /home/paul/cilium/bpf/lib/maps.h:8:
    In file included from /home/paul/cilium/bpf/lib/ipv6.h:9:
    /home/paul/cilium/bpf/lib/dbg.h:155:13: error: cast to smaller integer type '__u32' (aka 'unsigned int') from '__u32 *' (aka 'unsigned int *') [-Werror,-Wpointer-to-int-cast]
                    .source = EVENT_SOURCE,
                              ^~~~~~~~~~~~
    bpf_lxc.c:12:22: note: expanded from macro 'EVENT_SOURCE'
    #define EVENT_SOURCE LXC_ID
                         ^~~~~~
    /home/paul/cilium/bpf/lxc_config.h:14:16: note: expanded from macro 'LXC_ID'
    #define LXC_ID fetch_u32(LXC_ID)
                   ^~~~~~~~~~~~~~~~~
    /home/paul/cilium/bpf/lib/static_data.h:11:22: note: expanded from macro 'fetch_u32'
    #define fetch_u32(x) __fetch(x)
                         ^~~~~~~~~~
    /home/paul/cilium/bpf/include/bpf/ctx/../compiler.h:46:22: note: expanded from macro '__fetch'
    # define __fetch(X)             (__u32)(&(X))
                                    ^~~~~~~~~~~~~
Fix it by casting to __u64 first.

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

borkmann committed:
bpf: provide stddef/stdbool/stdio and avoid depending on compiler's

Loosly following,

[ upstream commit f1749d0303046f68f2f4a77584bc15b54c8edf58 ]

Remove the dependency on the system's definition of bool, true,
false and NULL. By dropping stdio/stdint/stddef. If we don't
do this we have two problems. First, we diverge further from
master and need additional mechanical updates. Second we have
to do more work on building clang side. This seems like the
best route.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

borkmann committed:
bpf, cilium: manually probe for v1/v2 BPF CPU availability

[ upstream commit 809bfff5678996dc0affc98874b0de3c6294e314 ]

There are a couple of downsides using -mcpu=probe right now with the latest
{clang,llc}-11. On latest kernels, -mcpu=probe will select v3 CPU which for
Cilium fails to pass the verifier due to buggy alu32 logic today. I've tried
running with -mcpu=probe -mattr=-alu32 but that doesn't work either at the
moment.

Another downside in LLVM's -mcpu=probe implementation is that anything else
than underlying x86_64 will just always fall back to v1 CPU, sigh:

  StringRef sys::detail::getHostCPUNameForBPF() {
  #if !defined(__linux__) || !defined(__x86_64__)
    return "generic";
  #else
    [...]
  #endif
  }

Given Cilium is in process of being supported on arm64, this is suboptimal.

Therefore, implement a manual probe and select between v1/v2 CPU from the
Cilium agent.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
runtime: cilium-runtime bump to llvm-10 and clang-10

Backporting 'bpf: Preserve source identity for hairpin via stack' is
currently causing a verifier error because of clang-7 generates the
following code,

 1530: (07) r1 += 8"
 1531: (47) r2 |= 3840"
 1532: (63) *(u32 *)(r1 +0) = r2"

Resulting in this verifier error on bpf-next,

 "dereference of modified ctx ptr R1 off=8 disallowed" subsys=datapath-loader

This is because r1 is the ctx pointer above and line 1530 adds to it and
then later does the store in 1532.

However, later llvm generates code without this,

    r3 |= r2
    *(u32 *)(r1 + 8) = r3

where the store uses the 'r1 + 8'.

To clean this up without resorting to hacks in the C code, there was
no clean way to convince clang-7 to generate code similar to clang-10
that I found, simply update runtime to use clang-10.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

pchaigno reviewed (approved):
:+1: for my changes.

jrfastab committed:
cilium: build bpf without -g option

[ upstream commit 8558a84b0e896412ffe4bf3e20c66a0c71cd340e ]

Using -g creates BTF errors. Quoting Daniel's upstream commit,

 bpf: only compile alignchecker with debug information

 iproute2's BTF parser otherwise breaks down on newer clang/llvm's
 generated BTF, therefore don't generate it for object files we are
 later on loading otherwise the daemon will not be able to start up.
 We need to find a better fix at some point.

Details are different in this patch due to 1.7 branch but it
is for the same reason.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

borkmann commented on bpf/init.sh:
```diff
@@ -275,7 +275,7 @@ function bpf_compile()
 	      $EXTRA_OPTS						\
 	      -c $LIB/$IN -o - |					\
-	llc -march=bpf -mcpu=probe -mattr=dwarfris -filetype=$TYPE -o $OUT
+	llc -march=bpf -mcpu=v2 -mattr=dwarfris -filetype=$TYPE -o $OUT
```
Problem is that this will fail verifier on older kernels where only v1 is supported. I presume that is why you saw the most recent verifier jmp opcode issue.

jrfastab replied:
right, I guess we need the same logic as master here...

jrfastab replied:
Just pulled in this,
```
 bpf, cilium: manually probe for v1/v2 BPF CPU availability

[ upstream commit 809bfff ]
```
WDYT?

borkmann replied:
That's generally okay. Do we have the probe manager as well in 1.7? Meaning, we run both, the bpftool probes as well as these https://github.com/cilium/cilium/tree/pr/v1.7-backport-2020-04-24/bpf/probes in the 1.7 tree, right?
